lib/prestashopper/api.rb

```diff
@@ -22,33 +22,28 @@ def initialize(url, key)
     # List resources that the API key can access
     # @return [Array<Symbol>] list of resources the API can access
     def resources
-      xml = @resources_res.get.body
-      xml_doc = Nokogiri::XML xml
-      nodes = xml_doc.xpath '/prestashop/api/*'
-      resources_list = []
-      nodes.each{|n| resources_list << n.name.to_sym}
-      return resources_list
+      @resources ||= Nokogiri::XML(@resources_res.get.body).xpath('/prestashop/api/*').collect { |resource| resource.name.to_sym }
     end
 
-    # Get all products data
-    # @return [Array<Hash>] list of products. Each product is represented by a hash with all its attributes.
-    def get_products
-      # /api/products returns XML with the IDs of each individual product
-      xml_products = @resources_res['products'].get.body
-      xml_products_doc = Nokogiri::XML xml_products
-      products_nodes = xml_products_doc.xpath '/prestashop/products/*/@id'
-      ids_list = []
-      products_nodes.each{|n| ids_list << n.value}
-
-      # GET each individual product to get the whole data
-      products = []
-      ids_list.each do |id|
-        xml_product = @resources_res["products/#{id}"].get.body
-        product = Product.xml2hash xml_product
-        products << product
+    def method_missing(method, *args, &block)
+      if method.to_s.starts_with?('get_')
+        resource = method.to_s.sub(/^get_/,'').pluralize.to_sym
+        raise 'You do not have access to this resource' unless resources.include?(resource)
+        if args.any?
+          ids = args.first.is_a?(Array) ? args.first : args
+          objects = ids.collect do |resource_id|
+            Hash.from_xml(Nokogiri::XML(@resources_res[resource][resource_id].get.body).remove_namespaces!.xpath("/prestashop/#{resource.to_s.singularize}").to_s).values
+          end.flatten.collect do |resource_hash|
+            resource_class_name = resource.to_s.classify
+            resource_class = "Prestashopper::#{resource_class_name}".safe_constantize || Prestashopper.const_set(resource_class_name, Class.new(OpenStruct))
```

Comment:
resource classes are generated on the fly, this will prevent the gem from needing an update when available resources are added.

Additionally, this code will still allow for a the class to be defined in case a custom parser is required or additional methods are added.